### ISSUE TITLE
Order of multiple ORDER BY statements

### ISSUE THREAD

**Issue by henrinormak** [typeorm/typeorm#2192]
**Issue type:**

[x] question
[ ] bug report
[x] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`

**TypeORM version:**

[ ] `latest`
[ ] `@next`
[x] `0.1.21`

**Steps to reproduce or a small repository showing the problem:**

When using multiple ORDER BY statements (via `addOrderBy` on `SelectQueryBuilder`), the order of those clauses in the raw SQL query is not guaranteed. This is possibly due to the way the ORDER BY clauses are mapped [here](https://github.com/typeorm/typeorm/blob/1d1457dc860e9d6bcdc9f606f4e9867ae05d6cb3/src/query-builder/SelectQueryBuilder.ts#L1506).

In short, although it seems to work, it might be beneficial to migrate the underlying storage mechanism of the ORDER BY statements to a Map, as that holds the items in order of insertion.

**Comment by pleerock** [comment]
yeah technically order is not guaranteed, but practically it always is. And I don't think it will be ever changed, most likely I believe that spec will change in order to guarantee keys in insertion order. Using map may become overhead for users who still use older ES versions and overall Map syntax is not as familiar for them as literal object syntax. So, I think it shall stay as it is, until there is a real issue with this approach.

**Comment by HELSIS666** [comment]
I ran into a similar problem. I have one endpoint for processing several optional filters (conditionally this is a search by: id, geolocation, name and other properties of the entity). When select game entities with many-to-many relationship to user table (sorting is not used) by any object properties, except for id, I get the correct users array order, and when choosing specifically by id, the user array is displayed in a strange order. This behavior surprised me very much. All these filters must select an array of objects or an array with one object (when selecting by identifier) ​​using the following construction:
```ts
@Injectable()
class FooClass {
  constructor(
    @InjectRepository(GameEntity)
    private readonly gameRepository: Repository<GameEntity>,
  ) {}

  public async selectAll(
    where: FindManyOptions['where'],
    relations: FindManyOptions['relations'] = ['sport', 'users'],
  ): Promise<GameEntity[]> {
    const options = { where, relations, cache: this.TTL };
    return await this.gameRepository.find(options).catch(() => {
      throw new NotFoundException('Games not found');
    });
  }
}
```
